src/daft-plan/src/optimization/rules/push_down_projection.rs

```diff
@@ -0,0 +1,425 @@
+                    if left_combined_dependencies.len() < left_upstream_names.len() {
+                        let pushdown_column_exprs = left_combined_dependencies
+                            .into_iter()
+                            .map(|s| Expr::Column(s.into()))
+                            .collect::<Vec<_>>();
+                        let new_project: LogicalPlan = Project::try_new(
+                            join.input.clone(),
+                            pushdown_column_exprs,
+                            Default::default(),
+                        )?
+                        .into();
+                        Some(new_project.into())
+                    } else {
+                        None
+                    }
+                };
+
+                let maybe_new_right_upstream: Option<Arc<LogicalPlan>> = {
+                    if right_combined_dependencies.len() < right_upstream_names.len() {
+                        let pushdown_column_exprs = right_combined_dependencies
+                            .into_iter()
+                            .map(|s| Expr::Column(s.into()))
+                            .collect::<Vec<_>>();
+                        let new_project: LogicalPlan = Project::try_new(
+                            join.right.clone(),
+                            pushdown_column_exprs,
+                            Default::default(),
+                        )?
+                        .into();
+                        Some(new_project.into())
+                    } else {
+                        None
+                    }
+                };
+
+                // If either pushdown is possible, create a new Join node.
+                if maybe_new_left_upstream
+                    .as_ref()
+                    .or(maybe_new_right_upstream.as_ref())
+                    .is_some()
+                {
+                    let new_left_upstream = maybe_new_left_upstream.unwrap_or(join.input.clone());
+                    let new_right_upstream = maybe_new_right_upstream.unwrap_or(join.right.clone());
+                    let new_join =
+                        upstream_plan.with_new_children(&[new_left_upstream, new_right_upstream]);
+                    let new_plan = plan.with_new_children(&[new_join]);
+                    // Retry optimization now that the upstream node is different.
+                    let new_plan = self
+                        .try_optimize(new_plan.clone())?
+                        .or(Transformed::Yes(new_plan))
+                        .unwrap()
+                        .clone();
+                    Ok(Transformed::Yes(new_plan))
+                } else {
+                    Ok(Transformed::No(plan))
+                }
+            }
+            LogicalPlan::Distinct(_) => {
+                // Cannot push down past a Distinct,
+                // since Distinct implicitly requires all parent columns.
+                Ok(Transformed::No(plan))
+            }
```

Comment:
I've been adding a catch-all `_ => Ok(Transformed::No(plan))` match arm so we don't need to update a bunch of unrelated optimization rules whenever we add a new logical node, and optimization rules need to explicitly opt-in to considering a logical op (partial pattern matching).

Are you thinking that all logical ops should have to explicitly opt-out for `PushDownProjection` since column pruning is a pretty universal optimization?

Reply:
That makes sense to me! 👍